### DISCUSSION TITLE
Subclassing existing sampling algorithms / kernels?

### DISCUSSION THREAD

**Discussion by mhinne** [blackjax-devs/blackjax#410] [Q&A] [answered]
Hi all, 

I've been toying around with Blackjax for a bit and I'm loving the experience. Boy are things fast :-)

Now that I've built up some experience with the existing samplers, I'm working on creating some extensions of these samplers for my specific needs. For instance, I've used the suggestions in #275 to construct a Gibbs kernel that combines elliptical slice sampling and RMH for a Gaussian process + its hyperparameters. This works very well, but now that I want to move on from ad-hoc fixes and a bit more clean design, I run into two (closely related) issues:

1. I wanted to subclass the ESS sampler to exploit some independencies in my specific likelihood. However, I cannot import from `blackjax.util`. I can of course copy-paste the `generate_gaussian_noise` and `linear_map` functions from the source code, but this seems clunky for an otherwise straightforward procedure.
2. I'm trying to use the Gibbs sampler from #275 within SMC, to faster explore the whole space of the Gaussian process hyperparameters (cf. [Svensson et al., 2015](https://arxiv.org/abs/1502.01908), except that I am using a non-conjugate likelihood), but this requires that I replace my ad-hoc Gibbs step function (as suggested in #275) with a proper Gibbs `MCMCSamplingAlgorithm` object. However, I cannot import this type from `blackjax.base`.

Am I missing an obvious way of importing these objects?

Thanks for your input :-)

**Comment by rlouf** [comment, marked as answer]
Hey, 

1. Could you please paste here the import that generates an error? `generate_gaussian_noise` is imported in a couple places in the lirbary using `from blackjax.util import generate_gaussian_noise`.
2. You can create a new SMC kernel by calling `from blackjax.smc.base import kernel` ([here in the code](https://github.com/blackjax-devs/blackjax/blob/2ac2fb4a1fad3789cfa840d2532865cbe2895bb8/blackjax/smc/base.py#L42)). This constructor requires two functions: 
   - a `mcmc_kernel_factory` (you can see what this corresponds to [here](https://github.com/blackjax-devs/blackjax/blob/2ac2fb4a1fad3789cfa840d2532865cbe2895bb8/blackjax/kernels.py#L125)), which is just a function that returns a sampling step function given a `logprob` function.
   - a `mcmc_kernel_generator`, which given a position in the sample space returns the state that your sampling step function takes as an input. This is important for HMC, for instance: SMC takes positions as inputs, but HMC needs extra information.

This means that you don't have to subclass `MCMCSamplingAlgorithm` at all, and can just work with pure step functions like the Gibbs function you constructed! You are touching on some minor clunkiness with the SMC sampler that I'm trying to solve here https://github.com/blackjax-devs/blackjax/pull/279, and the lack of documentation :grimacing: but you should totally be able to do it. This is a perfectly reasonable use case, thus it shouldn't be too difficult to do; if it turns out to be we'll need to refactor the internals a little.

If that answer is not enough, please share a simple example and we'll walk through it together.

**Reply by mhinne** [reply]
Hey,

1. Here is the error I get when importing from `blackjax.util`:

```
import blackjax
from blackjax.util import generate_gaussian_noise`

     1 import blackjax
----> 2 from blackjax.util import generate_gaussian_noise

ModuleNotFoundError: No module named 'blackjax.util'
```

Just importing `blackjax.util` gives the same result, and `from blackjax import util` yields

```
from blackjax import util
ImportError                               Traceback (most recent call last)

[<ipython-input-4-db1738faee1b>](https://localhost:8080/#) in <module>
      1 import blackjax
----> 2 from blackjax import util

ImportError: cannot import name 'util' from 'blackjax' (/usr/local/lib/python3.7/dist-packages/blackjax/__init__.py)
```

2. Ah that sounds great and exactly how I hoped it ought to work - I've been making things too complicated :-) I'll give it another go and will let you know my progress.

Thanks for your help!

**Reply by rlouf** [reply]
For 1., are you sure you are working with the code on the `main` branch? I have no problem running the import locally.

**Reply by mhinne** [reply]
(continuing with 1) Hmm as far as I am aware yes, I'm working in Google colab and what I do is 

`pip install blackjax`
`import blackjax`
`from blackjax.util import generate_gaussian_noise` 

If I let colab autocomplete my options after `from blackjax.`, I only have these fields/modules accessible:

- adaptation
- base
- diagnostics
- kernels
- mcmc
- optimizers
- progress_bar
- sgmcmc
- smc
- types
- vi
- _version

Am I installing blackjax the wrong way?

**Reply by rlouf** [reply]
The current version on PyPi is different from what's on the `main` branch currently (we're refactoring for v1). Try:

```
pip install git+https://github.com/blackjax-devs/blackjax 
```

We'll soon cut nightly release to make this easier.

**Reply by mhinne** [reply]
I feel very silly now, that fixed it completely... thanks a lot for the help!

**Reply by rlouf** [reply]
No problem. Things work differently on the `main` branch, and we’re happy to get feedback if you feel it’s worse than on the latest release.

**Reply by rlouf** [reply]
But the way you can now use the nightly releases to work with the latest changes:

```
pip install blackjax-nightly
```

**Reply by mhinne** [reply]
Great, I'll keep an eye out! Looking forward to the v1 release as well :-)

**Reply by rlouf** [reply]
Me too 😭

**Reply by mhinne** [reply]
Just to follow up for others interested in something similar, we (my PhD student did the hard work) got the SMC approach with our Gibbs kernel working. We did simplify things so I hope that is all correct, but the results seem very sensible. In the end, we did instantiate our Gibbs kernel as a new `MCMCSamplingAlgorithm`. With that, it was simply a matter of calling the `adaptive_tempered_smc` sampler as in the example, replacing only NUTS/HMC with our Gibbs kernel. It works and is hella fast :-)

**Reply by rlouf** [reply]
Awesome! Is that Metropolis-within-Gibbs-within-SMC then? 😁

**Reply by mhinne** [reply]
Even better! It is (Elliptical Slice Sampling and Metropolis)-within-Gibbs-within-SMC :-)

**Reply by Hesterhuijsdens** [reply]
Hi @rlouf ,

I am working together with @mhinne to combine the SMC approach with our Gibbs kernel. As mentioned above, we thought we could just create an instance of `MCMCSamplingAlgorithm` (or, when using the new Blackjax version, an init function and a step function) for our Gibbs kernel, and feed this into the `tempered_smc` sampler. 

However, we now noticed that, within the [tempered SMC script](https://github.com/blackjax-devs/blackjax/blob/main/blackjax/smc/tempered.py), the `tempered_logposterior_fn` is defined and given to the (Gibbs) kernel. However, since we want to use Elliptical Slice Sampling within the Gibbs kernel, we would require a tempered log **likelihood** function instead.

We have now adapted our local version of the tempered script so that it feeds a `tempered_loglikelihood_fn` into the Gibbs kernel, but I was wondering if there is a nicer way to achieve this, without having to adapt the Blackjax scripts? 🙂

**Reply by maxhinne** [reply]
Would it be possible to have the SMC temperature parameter (`lmbda`) be part of the `mcmc_init_fn` or `mcmc_step_fn`, in lines 128 and 131 in https://github.com/blackjax-devs/blackjax/blob/main/blackjax/smc/tempered.py? That way, any custom kernel (that doesn't take the `tempered_logposterior_fn` as input, like our Gibbs kernel) can use the temperature internally :-)

**Reply by rlouf** [reply]
That's a fair point, please open an issue that describes your problem.

**Comment by AdrienCorenflos** [comment]
By the way, have you considered using [this](https://rss.onlinelibrary.wiley.com/doi/full/10.1111/rssb.12269) instead of ESS? It's been proven to work better on a number of applications. We have an implementation in BlackJAX. See here for example
https://blackjax-devs.github.io/blackjax/examples/LogisticRegressionWithLatentGaussianSampler.html

**Reply by mhinne** [reply]
Hmm I had not, mostly out of ignorance. It seems this approach can not only outperform ESS for the latent Gaussians themselves, but also for their hyperparameters which is a big challenge in our project. So I'll definitely give this a look, thanks for the tip! :-)

**Reply by AdrienCorenflos** [reply]
Check their "z-" sampler.

**Reply by mhinne** [reply]
That looks spot-on indeed! Not to sound too lazy, but this is not part of BlackJAX (yet), right, only the marginal sampler? :)
